### PR TITLE
Add multi texture support for sprites

### LINKED ISSUE #3233: Multiple textures on sprites
In order to do multiple passes with different shaders and/or tags we need to be able to add many material and texture pairs to a sprite or a spine.

*Use case:*
The dynamic light engine for Cohesion needs to be able to draw the same sprite with different shaders and textures to different render targets Currently this is solved by having multiple sprite components on each game object. This works "ok" with simple sprites (allthough syncing animations could be a problem), but with spines it means having two spine animations to keep in sync, and the extra overhead of having two spines to manage. It simply isn't a scalable workflow.
2016-05-30 12:47:19 (ingela.garneij)
Needed before Cohesion playtests (October?)
2016-09-21 12:10:22 (Sara.Cederberg)
[~sven.andersson] How does this relate to (LINK REMOVED)
2017-06-12 10:49:51 (Fredrik.Malmer)
Would like to add that this is also needed on models.
2017-09-14 16:57:18 (Fredrik.Malmer)
I would recommend taking a holistic view when designing this and also consider the following key missing related features:

* Overrides of material properties on the sprite/model/particle/gui (setting the sprite atlas today is just a special case of this).
* Default texture file assignments in materials
* Adding render state settings to materials

Doing this feature without considering these things would likely lead to a design and a file format that will need to be redone soon in the future.
2017-10-03 14:38:08 (Fredrik.Malmer)
Also to consider: Right now models can only take in one object with one material. This is very limiting. What we really need for models is to be able to contain several objects with different materials. In other words support for many materials per model, preferably through the use of material ids. See: DEF-2910

Any design needs to account for this very probable future.

### PR DESCRIPTION
**Pull request by JCash**
This feature lets you assigne multiple atlases to a sprite, enabling you to do effects like normal mapping etc.

By adding multiple `texcoord` vertex attributes to your material, and also adding multiple texture samplers, you will enable this feature for the sprites using your material.

Fixes https://github.com/defold/defold/issues/3233

## PR checklist

* [ ] Code
	* [ ] Add engine and/or editor unit tests.
	* [x] New and changed code follows the overall code style of existing code
	* [x] Add comments where needed
* [ ] Documentation
	* [ ] Make sure that API documentation is updated in code comments
	* [ ] Make sure that manuals are updated (in github.com/defold/doc)
* [ ] Prepare pull request and affected issue for automatic release notes generator
	* [ ] Pull request - Write a message that explains what this pull request does. What was the problem? How was it solved? What are the changes to APIs or the new APIs introduced? This message will be used in the generated release notes. Make sure it is well written and understandable for a user of Defold.
	* [ ] Pull request - Write a pull request title that in a sentence summarises what the pull request does. Do not include "Issue-1234 ..." in the title. This text will be used in the generated release notes.
	* [ ] Pull request - Link the pull request to the issue(s) it is closing. Use on of the [approved closing keywords](https://docs.github.com/en/issues/tracking-your-work-with-issues/linking-a-pull-request-to-an-issue).
	* [ ] Affected issue - Assign the issue to a project. Do not assign the pull request to a project if there is an issue which the pull request closes.
	* [ ] Affected issue - Assign the "breaking change" label to the issue if introducing a breaking change.
	* [ ] Affected issue - Assign the "skip release notes" is the issue should not be included in the generated release notes.

----------

Example of a well written PR description:

1. Start with the user facing changes. This will end up in the release notes.
1. Add one of the GitHub approved closing keywords
1. Optionally also add the technical changes made. This is information that might help the reviewer. It will not show up in the release notes. Technical changes are identified by a line starting with one of these:
   1. `### Technical changes` 
   1. `Technical changes:`
   2. `Technical notes:`

```
There was a anomaly in the carbon chroniton propeller, introduced in version 8.10.2. This fix will make sure to reset the phaser collector on application startup.

Fixes #1234

### Technical changes
* Pay special attention to line 23 of phaser_collector.clj as it contains some interesting optimizations
* The propeller code was not taking into account a negative phase.
```

### CODE REVIEW

File: engine/gamesys/src/gamesys/resources/res_textureset.cpp [hunk 1]
```diff
@@ -41,66 +42,75 @@ namespace dmGameSystem
...
+        if (r != dmResource::RESULT_OK)
         {
-            // Get path for texture
-            r = dmResource::GetPath(factory, texture_res, &tile_set->m_TexturePath);
-            if (r != dmResource::RESULT_OK) {
-                return r;
-            }
+            dmDDF::FreeMessage(texture_set_ddf);
+            return r;
+        }
```
Comment (JCash) [inline engine/gamesys/src/gamesys/resources/res_textureset.cpp:49]: I prefer early out

File: engine/gamesys/src/gamesys/resources/res_textureset.cpp [hunk 2]
```diff
@@ -41,66 +42,75 @@ namespace dmGameSystem
...
         }
+
+        // This is a mapping from the single frame image names (animation ids) to the frame number
+        uint32_t n_image_name_hashes = texture_set_ddf->m_ImageNameHashes.m_Count;
+        tile_set->m_FrameIds.SetCapacity(dmMath::Max(1U, (n_image_name_hashes*2)/3), n_image_name_hashes);
+        for (uint32_t i = 0; i < n_image_name_hashes; ++i)
+        {
+            dmhash_t h = texture_set_ddf->m_ImageNameHashes[i];
+            tile_set->m_FrameIds.Put(h, i);
+        }
```
Comment (JCash) [inline engine/gamesys/src/gamesys/resources/res_textureset.cpp:112]: This part is new. We keep a lookup table from animation name (E.g. "foo" from "path/foo.png") into the frame index.
This allows us to lookup images between atlases.

File: engine/graphics/src/vulkan/graphics_vulkan.cpp [hunk 3]
```diff
@@ -2135,17 +2149,9 @@ namespace dmGraphics
...
-        uint16_t stream_count = vertex_declaration->m_StreamCount;
-        for (int i = 0; i < stream_count; ++i)
-        {
-            VertexDeclaration::Stream& stream = vertex_declaration->m_Streams[i];
-            dmHashUpdateBuffer32(state, &stream.m_NameHash, sizeof(stream.m_NameHash));
-            dmHashUpdateBuffer32(state, &stream.m_Location, sizeof(stream.m_Location));
-            dmHashUpdateBuffer32(state, &stream.m_Offset, sizeof(stream.m_Offset));
-            dmHashUpdateBuffer32(state, &stream.m_Format, sizeof(stream.m_Format));
-        }
+        dmHashUpdateBuffer32(state, vertex_declaration->m_Streams, sizeof(VertexDeclaration::Stream) * vertex_declaration->m_StreamCount);
```
Comment (JCash) [inline engine/graphics/src/vulkan/graphics_vulkan.cpp:2154]: No need to loop over each stream+attribute.

File: scripts/unpack_ddf.py [hunk 4]
```diff
@@ -57,8 +58,10 @@
 BUILDERS['.fontc']          = render.font_ddf_pb2.FontMap
 BUILDERS['.glyph_bankc']    = render.font_ddf_pb2.GlyphBank
 BUILDERS['.particlefxc']    = particle.particle_ddf_pb2.ParticleFX
+BUILDERS['.spritec']        = gamesys.sprite_ddf_pb2.SpriteDesc
 
 proto_type_to_string_map = {}
+proto_type_to_string_map[google.protobuf.descriptor.FieldDescriptor.TYPE_BOOL]    = 'TYPE_BOOL'
```
Comment (JCash) [inline scripts/unpack_ddf.py:64]: Add support for enums...

File: scripts/unpack_ddf.py [hunk 5]
```diff
@@ -93,8 +96,8 @@ def get_descriptor_type(descriptor):
 
 def print_descriptor_default(descriptor, data):
     if descriptor.type == descriptor.TYPE_ENUM:
-        enum_name = descriptor.enum_type.values[data].name
-        print(descriptor.name, ":", enum_name, get_descriptor_type(descriptor))
+        e = descriptor.enum_type.values_by_number.get(data)
+        print(descriptor.name, ":", e.name, get_descriptor_type(descriptor))
```
Comment (JCash) [inline scripts/unpack_ddf.py:100]: .. and properly getting the correct value.

File: engine/gamesys/proto/gamesys/atlas_ddf.proto [hunk 6]
```diff
@@ -25,6 +25,7 @@ message AtlasAnimation
     optional uint32 flip_vertical   = 6 [default = 0];
 }
 
+// Editor format. Engine format is iun texture_set_ddf.proto
```
Comment (matgis) [inline engine/gamesys/proto/gamesys/atlas_ddf.proto:28]: Typo: `iun`

**Review by matgis — CHANGES_REQUESTED** [review]
Editor & file format changes look good!

However, I think something might have broken when bundling with Bob. I get this error when I try to bundle a fresh project created from the Side Scroller Tutorial template:
```
/game.project
	File /Users/mats/Defold/Side scroller tutorial2/build/default/background/background.atlas does not exists
```

File: engine/gamesys/src/gamesys/components/comp_sprite.cpp [hunk 7]
```diff
@@ -475,35 +687,46 @@ namespace dmGameSystem
...
+    }
 
-            vertices_write_ptr += info->m_ValueByteSize;
+    static void EnsureSize(dmArray<float>& array, uint32_t size)
+    {
+        if (array.Capacity() < size) {
+            array.OffsetCapacity(size - array.Capacity());
         }
+        array.SetSize(size);
     }
```
Comment (Jhonnyg) [inline engine/gamesys/src/gamesys/components/comp_sprite.cpp:724]: Maybe this should be a first class function in dmarray, we use this pattern alot afaik

**Review by Jhonnyg — DISMISSED** [review]
LGTM

**Pull request merged**